File: backend/app/schemas/artifacts.py
from datetime import datetime
from typing import Any, Literal
import uuid

from pydantic import BaseModel, Field


ArtifactLifecycleStatusLiteral = Literal["draft", "reviewed", "submitted", "archived"]
ArtifactTypeLiteral = Literal["tailored_resume", "cover_letter"]


class ArtifactDraftCreate(BaseModel):
    workspace_id: uuid.UUID
    role_id: uuid.UUID | None = None
    opportunity_id: uuid.UUID | None = None
    application_id: uuid.UUID | None = None
    artifact_type: ArtifactTypeLiteral
    title: str = Field(min_length=1, max_length=255)
    content: str = Field(min_length=1)
    evaluation_id: uuid.UUID | None = None
    source_resume_version_id: uuid.UUID | None = None
    source_artifact_id: uuid.UUID | None = None
    change_summary: str | None = Field(default=None, max_length=5000)


class ArtifactDraftUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    content: str | None = Field(default=None, min_length=1)
    change_summary: str | None = Field(default=None, max_length=5000)


class ArtifactTraceabilityResponse(BaseModel):
    workspace_id: uuid.UUID
    role_id: uuid.UUID | None = None
    opportunity_id: uuid.UUID | None = None
    application_id: uuid.UUID | None = None
    evaluation_id: uuid.UUID | None = None
    source_resume_version_id: uuid.UUID | None = None
    source_artifact_id: uuid.UUID | None = None
    parent_artifact_id: uuid.UUID | None = None
    generation_warnings: list[str] = Field(default_factory=list)
    export_formats: list[str] = Field(default_factory=list)


class ArtifactResponse(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    application_id: uuid.UUID | None = None
    role_id: uuid.UUID | None = None
    opportunity_id: uuid.UUID | None = None
    artifact_type: str
    lifecycle_status: ArtifactLifecycleStatusLiteral
    version_number: int
    title: str
    content: str
    reviewed_at: datetime | None = None
    submitted_at: datetime | None = None
    archived_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    traceability: ArtifactTraceabilityResponse
    available_transitions: list[ArtifactLifecycleStatusLiteral] = Field(default_factory=list)
    new_version_created: bool = False
    source_submitted_artifact_id: uuid.UUID | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)

File: frontend/src/types/artifacts.ts
export type ArtifactLifecycleStatus =
  | "draft"
  | "reviewed"
  | "submitted"
  | "archived";

export type ArtifactType = "tailored_resume" | "cover_letter";

export interface ArtifactTraceability {
  workspace_id: string;
  role_id: string | null;
  opportunity_id: string | null;
  application_id: string | null;
  evaluation_id: string | null;
  source_resume_version_id: string | null;
  source_artifact_id: string | null;
  parent_artifact_id: string | null;
  generation_warnings: string[];
  export_formats: string[];
}

export interface ArtifactRecord {
  id: string;
  workspace_id: string;
  application_id: string | null;
  role_id: string | null;
  opportunity_id: string | null;
  artifact_type: ArtifactType | string;
  lifecycle_status: ArtifactLifecycleStatus;
  version_number: number;
  title: string;
  content: string;
  reviewed_at: string | null;
  submitted_at: string | null;
  archived_at: string | null;
  created_at: string;
  updated_at: string;
  traceability: ArtifactTraceability;
  available_transitions: ArtifactLifecycleStatus[];
  new_version_created: boolean;
  source_submitted_artifact_id: string | null;
  metadata: {
    revision_id?: string | null;
    change_summary?: string | null;
  };
}

export interface ArtifactDraftUpdatePayload {
  title?: string;
  content?: string;
  change_summary?: string | null;
}
